README
The updater script
==================

When this script is run, it downloads list of packages and upgrades the system
based on the list.

It can handle lists of packages individual to the given device and individual
encrypted packages.

The URL of the repository and of the package list is hardcoded at the top of
the script.

The repository layout
---------------------

Assume that `$ID` is the serial number of the device in question. The
`$BASE_URL` is the top level of the repository structure.

The list for the device is located as `$BASE_URL/lists/$ID`. If that one does
not exist, the updater script takes a generic one, at
`$BASE_URL/lists/generic`.

An unencrypted package is located at address
`$BASE_URL/packages/<packagename>-<packageversion>.ipk`. An encrypted one is at
`$BASE_URL/packages/<packagename>-<packageversion>-$ID.ipk`.

Note that the repository may contain packages in multiple versions.

The package list format
-----------------------

Each line represents one package. The format of the line is:

  <package-name>	<version>	<flags>

The columns are separated by one tab. The version is the version that should be
installed on the device.

Flags represent special properties of the package. Each one is an on-off
switch. Each flag is single letter, all of them are concatenated together.

E::
  The package is encrypted and individual for the given device.
R::
  Instead of installing the package, the package should be removed from the
  system. The version should be `-` in this case.
F::
  Force re-installation even if it is in current version.
C::
  Do a cooldown sleep after installing or removing this package.
U::
  Request a restart of updater. After installing the package, start updater
  from the beginning.
I::
  Update the package only if already installed. Do not install if not installed
  yet. Used for upgrading packages that need exact version if they are installed,
  but not crucial for working of the device, like kernel modules.
X::
  The name is not a name of package directly, but a regular expression. All packages
  matching it will be included with the same flags.
L::
  The name is not a package, but a list to generate in addition to the current one.
  The list will be read from the `<package>` file and written to
  name without any suffix (eg. the `<package>` is stripped after first dot).
  Packages in the current list will be omited from the additional one.
B::
  This package requires reboot-installed updates. See below.
G::
  Ignore the package. That is, don't install or update it, but don't remove it either.
D::
  Deprecate old versions. Remove the package if it has different version than
  the one listed.
P::
  Passthrough flag for the generator. Ignored in the client.
<number>::
  Packages with lower number are considered sooner when planning the order.
  Packages without number are left as last.

Dependencies
------------

Dependencies are not considered on the device. The list is processed from top
to bottom, one line at a time. If one package depends on another, it should be
listed in the list above the another.

Reboot-installed updates
------------------------

In case the list of packages that need to be updated contains some with the `B`
flag, the packages are just downloaded and prepared and a reboot is scheduled.
After the reboot, the updates are installed and another reboot is done. This is
to accomodate critical services that can't be restarted while the system is
running (like `procd`).

This works only in the primary list, not in the additional ones.

Example
-------

This is an example of the package list that includes ucollect with its all
dependencies and removes nuci.

  libgcc	4.6-linaro-1	
  libc	0.9.33.2-1	
  libpcap	1.1.1-2	
  libpthread	0.9.33.2-1	
  librt	0.9.33.2-1	
  libopenssl	1.0.1e-1	
  zlib	1.2.7-1
  libuci	2013-05-02.1-1	
  ucollect	14	E
  nuci	-	R

Overriding things for testing
-----------------------------

The `updater.sh` script autodetects several things when it starts. It is
possible to override some of them for testing purposes. It is not expected to
be used by ordinary users.

If you want to override anything, set corresponding uci option. If the option
is not set, the default (autodetection) is used. The config section is not even
present in the default config.

 * `updater.override.branch`: The branch used is selected according to the atsha
   serial number. The setting for which serial is in which branch is on the server.
   This allows to override the server configuration and force given branch, like
   `testing` or `rc`. Also, it is possible to set this to a serial number to fake
   another device.
 * `updater.override.revision`: The hardware revision used. This is the first half
   of the serial number (like `00000005`). It is little use currently, as all
   the existing revisions contain the same packages.
 * `updater.override.generation`: The firmware generation. The autodetection takes
   the major version of `turris-version` and appends a slash (eg. `2/`). If setting
   this, don't forget the slash.
 * `updater.override.base_url`: The url where the packages are downloaded from.
   The default is `https://api.turris.cz/updater-repo/`.
 * `updater.override.list_req_url`: The url where request for package list is sent to
   (including the revision, generation and ID).  This CGI script then returns bundle
   of package lists.
 * `updater.override.hash_url`: The directory where hash descriptions live.
 * `updater.override.disable`: If set to `true`, the updater won't do anything.
   Not recommended.

The generator script
====================

The generator script can generate the list of packages and the directory with
the package archives.

To run it, pass the base URL of the openWRT repository to generate the packages
from. Then, pass the list of desired packages to its standard input, one
package per line. A package name can be followed by flags. There's another flag
in addition to the ones in package list, `1`. If it is present, the operation
is placed at the top of the list (it is implemented for removal only now).

The generator tracks dependencies and includes them in the final list.

It is integrated into jenkins, so the basic list is regenerated automatically.
The list of packages is in the `list-desired` files.

The deployment script
=====================

There are currently 3 stages of stability of the OpenWRT repository. The first
(most unstable) is in the `test` branch. Whenever a new version is pushed,
Jenkinst builds the basic set of packages and generates updates for them. These
are installed on few devices close at hand.

The second stage is in the branch `master`. It acts mostly the same, except that:

 * All available packages are built.
 * The updates are installed on few other devices that are not necessarily as
   close at hand.

The last deployment is done by copying what was previously built for the
`master` branch, after enough testing. To ensure integration into the rest of
the deployment process and to protect against deploying untested versions
without due testing, the following needs to be done:

 1. The hash of the tested commit on `master` branch is entered into the
    `deploy` file and a Signed-Off-By line is added by whoever adds the hash.
    Note that this is the hash in the commit in openWRT repository, inserted
    into file in the `misc` repository.

 2. This commit is pushed into special branch `pending`.
 3. Another person pulls that branch and (after checking the changes brought by
    such revision) adds their own Signed-Off-By line by:

	git commit --amend -s

 4. After that, the checker force-pushes to the `pending` branch and overwrites
    the commit by its current version:

	git push -f

 5. A third person signs the commit by a PGP key trusted by the jenkins build:

	git pull
	git commit --amend -S<keyid> # Note the capital S

 6. Merges the modified commit into master and pushes.
 7. Launches the `Deploy` job in Jeknins (it must be started manually).

[NOTE]
The commit which hash is in the `deploy` file must still be on the HEAD of the
`master` branch. If a newer has been pushed in between, the deploy won't work,
because it just copies the already compiled files.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Support for `FilesSignature` field in packages. On mismatch it trigger
  reinstall.

### Fixed
- Subprocess call is now terminated way earlier thanks to `SIGCHLD` signal
  handling. This improves update time for any scripts spawning "daemon" processes
  that do not correctly redirect or close standard outputs.


## [69.1.3] - 2021-06-16
### Changed
- Error generated from Lua when URI is being finished now includes error message
  from downloader if it was download failure.


## [69.1.2] - 2021-05-28
### Fixed
- Immediate reboot not being performed when combined with replan that actually
  performs some changes in the system


## [69.1.1] - 2021-04-15
### Fixed
- error on transaction recovery created by updater before version 69.0.1 about
  `upgraded_packages` being nil


## [69.1.0] - 2021-04-15
### Changed
- Package requests solver no longer tries to maximize number of selected requests
  but rahter follows strictly rules of requests priority
- `Install` requests with same priority as `Unistall` are resolved first
- `Install` and `Unistall` requests without condition as resolved before requests
  with condition given they have same priority specified.

### Removed
- Error reported when request to install and unistall same package was specified


## [69.0.1] - 2021-03-17
### Fixed
- environment variable `PKG_UPGRADE` were always set to `1` for package `postinst`
  script no matter if it was package upgrade or not.

### Changed
- usage of `example.org` replaced with `application-test.turris.cz` in tests


## [69.0.0] - 2020-11-24
### Added
- extra argument `pkg_hash_required` for `Repository` command
- support for version limitation of `Install` and `Uninstall` requests in package
  name (such as `foo (>= 1.0.0)`).

### Changed
- Execution is now terminated when there is no hash for package being installed in
  repository index unless `pkg_hash_required` is set to `false` for that
  repository.

### Removed
- Extra option `version` for `Install` request. You should append version
  specifier to package name the same way as it is done for dependencies.

### Fixed
- `ROOT_DIR` not being defined for `hook_postinst` and `hook_reboot_required` on
  replan execution.


## [68.0.0] - 2020-11-06
### Added
- Queue messages for 'upgrade' and 'downgrade' now also prints current version in
  square brackets.

### Changed
- Default connection timeout for download was extended from one minute to ten
  minutes.

### Removed
- Possibility to reboot immediatelly after package installation. Reboot after
  update is deemed sufficient.

### Fixed
- Queue messages now state 'upgrade', 'downgrade' and 'reinstall' instead of
  original generic 'install'.


## [67.0.3] - 2020-10-21
### Changed
- pkgupdate's conf.lua now loads scripts with Full security level instead of Local
- libupdater is now versioned with release version (there is no API compatibility
  between versions)


## [67.0.2] - 2020-08-06
### Fixed
- Warning about cycles for packages providing and at the same time conflicting
  with some other package
- Configure script now checks if uthash is available


## [67.0.1.1] - 2020-07-01
### Changed
- Lunit submodule now points to new Github repository


## [67.0.1] - 2020-06-25
### Changed
- Package "reinstall" is now performed not only if `Version` field is different
  but also when `Architecture`, `LinkSignature`, `Depends`, `Conflicts` or
  `Provides` are different.
- Information about package changes planned to be performed are now printed with
  wording signaling if that is new package or already installed one and if that is
  update or downgrade or generic reinstall.


## [67.0] - 2020-06-23
### Added
- Warning for packages not verified against repository index because missing hash

### Changed
- Custom build system was replaced with autotools
- OpenSSL is now used to verify signatures instead of usign
- URI implementation no longer uses temporally files and passed instead everything
  in memory
- Download retries are removed, code now relies only on libcurl reconnection
- libb64 usage replaced with OpenSSL

### Fixed
- Memory leak on archive open error
- Various compilation warning
- Invalid error complaining about path being called on on URI of invalid scheme


## [66.0] - 2020-04-27
### Changed
- libarchive is now used to unpack packages instead of tar command
- rm -rf call is replaced with built in function
- call to find replaced with internal function implementation
- empty journal recovery is now not considered as fatal

### Fixed
- fix invalid sha256sum field name and that way hash verification

### Removed
- update_alternatives.sh script was removed


## [65.0] - 2020-02-20
### Added
- Extra argument `condition` for Install and Uninstall
- Integrated support for Alternatives of packages in transaction


## [65.0] - 2020-02-20
### Added
- Mode command to control updater's special run modes from configuration
- Package scripts now have in environment variable PKG_UPGRADE signaling if it is
  new installation or just upgrade.


## [63.2] - 2020-01-20
### Added
- Add URI to some error messages

### Changed
- Virtual packages are now packages without candidate and any existing candidate
  for them is ignored (and removed if installed)

### Fixed
- Multiple Provices not being applied on appropriate package and effectively
  ignored


## [63.1.4] - 2019-11-26
### Fixed
- Package block if it provides itself


## [63.1.3] - 2019-11-06
### Fixed
- Invalid stack allocation in subprocess


## [63.1.2] - 2019-09-04
### Added
- Alternatives updating hook shipped with pkgupdate


## [63.1.1] - 2019-08-26
### Fixed
- `update_alternatives.sh` not working when root wasn't current one

src/lib/md5.c
/* Functions to compute MD5 message digest of files or memory blocks.
   according to the definition of MD5 in RFC 1321 from April 1992.
   Copyright (C) 1995-1997, 1999-2001, 2005-2006, 2008-2016 Free Software
   Foundation, Inc.
   This file is part of the GNU C Library.

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 2, or (at your option) any
   later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, see <http://www.gnu.org/licenses/>.  */

/* Written by Ulrich Drepper <drepper@gnu.ai.mit.edu>, 1995.  */

//#include <config.h>

#if HAVE_OPENSSL_MD5
# define GL_OPENSSL_INLINE _GL_EXTERN_INLINE
#endif
#include "md5.h"

#include <stdalign.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>

#if USE_UNLOCKED_IO
# include "unlocked-io.h"
#endif

#include <endian.h>
#if __BYTE_ORDER == __BIG_ENDIAN
# define WORDS_BIGENDIAN 1
#endif

#ifdef _LIBC
/* We need to keep the namespace clean so define the MD5 function
   protected using leading __ .  */
# define md5_init_ctx __md5_init_ctx
# define md5_process_block __md5_process_block
# define md5_process_bytes __md5_process_bytes
# define md5_finish_ctx __md5_finish_ctx
# define md5_read_ctx __md5_read_ctx
# define md5_stream __md5_stream
# define md5_buffer __md5_buffer
#endif

#ifdef WORDS_BIGENDIAN
# define SWAP(n)                                                        \
    (((n) << 24) | (((n) & 0xff00) << 8) | (((n) >> 8) & 0xff00) | ((n) >> 24))
#else
# define SWAP(n) (n)
#endif

#define BLOCKSIZE 32768
#if BLOCKSIZE % 64 != 0
# error "invalid BLOCKSIZE"
#endif

#if ! HAVE_OPENSSL_MD5
/* This array contains the bytes used to pad the buffer to the next
   64-byte boundary.  (RFC 1321, 3.1: Step 1)  */
static const unsigned char fillbuf[64] = { 0x80, 0 /* , 0, 0, ...  */ };


/* Initialize structure containing state of computation.
   (RFC 1321, 3.3: Step 3)  */
void
md5_init_ctx (struct md5_ctx *ctx)
{
  ctx->A = 0x67452301;
  ctx->B = 0xefcdab89;
  ctx->C = 0x98badcfe;
  ctx->D = 0x10325476;

  ctx->total[0] = ctx->total[1] = 0;
  ctx->buflen = 0;
}

/* Copy the 4 byte value from v into the memory location pointed to by *cp,
   If your architecture allows unaligned access this is equivalent to
   * (uint32_t *) cp = v  */
static void
set_uint32 (char *cp, uint32_t v)
{
  memcpy (cp, &v, sizeof v);
}

/* Put result from CTX in first 16 bytes following RESBUF.  The result
   must be in little endian byte order.  */
void *
md5_read_ctx (const struct md5_ctx *ctx, void *resbuf)
{
  char *r = resbuf;
  set_uint32 (r + 0 * sizeof ctx->A, SWAP (ctx->A));
  set_uint32 (r + 1 * sizeof ctx->B, SWAP (ctx->B));
  set_uint32 (r + 2 * sizeof ctx->C, SWAP (ctx->C));
  set_uint32 (r + 3 * sizeof ctx->D, SWAP (ctx->D));

  return resbuf;
}

/* Process the remaining bytes in the internal buffer and the usual
   prolog according to the standard and write the result to RESBUF.  */
void *
md5_finish_ctx (struct md5_ctx *ctx, void *resbuf)
{
  /* Take yet unprocessed bytes into account.  */
  uint32_t bytes = ctx->buflen;
  size_t size = (bytes < 56) ? 64 / 4 : 64 * 2 / 4;

  /* Now count remaining bytes.  */
  ctx->total[0] += bytes;
  if (ctx->total[0] < bytes)
    ++ctx->total[1];

  /* Put the 64-bit file length in *bits* at the end of the buffer.  */
  ctx->buffer[size - 2] = SWAP (ctx->total[0] << 3);
  ctx->buffer[size - 1] = SWAP ((ctx->total[1] << 3) | (ctx->total[0] >> 29));

  memcpy (&((char *) ctx->buffer)[bytes], fillbuf, (size - 2) * 4 - bytes);

  /* Process last bytes.  */
  md5_process_block (ctx->buffer, size * 4, ctx);

  return md5_read_ctx (ctx, resbuf);
}
#endif

/* Compute MD5 message digest for bytes read from STREAM.  The
   resulting message digest number will be written into the 16 bytes
   beginning at RESBLOCK.  */
int
md5_stream (FILE *stream, void *resblock)
{
  struct md5_ctx ctx;
  size_t sum;

  char *buffer = malloc (BLOCKSIZE + 72);
  if (!buffer)
    return 1;

  /* Initialize the computation context.  */
  md5_init_ctx (&ctx);

  /* Iterate over full file contents.  */
  while (1)
    {
      /* We read the file in blocks of BLOCKSIZE bytes.  One call of the
         computation function processes the whole buffer so that with the
         next round of the loop another block can be read.  */
      size_t n;
      sum = 0;

      /* Read block.  Take care for partial reads.  */
      while (1)
        {
          n = fread (buffer + sum, 1, BLOCKSIZE - sum, stream);

          sum += n;

          if (sum == BLOCKSIZE)
            break;

          if (n == 0)
            {
              /* Check for the error flag IFF N == 0, so that we don't
                 exit the loop after a partial read due to e.g., EAGAIN
                 or EWOULDBLOCK.  */
              if (ferror (stream))
                {
                  free (buffer);
                  return 1;
                }
              goto process_partial_block;
            }

          /* We've read at least one byte, so ignore errors.  But always
             check for EOF, since feof may be true even though N > 0.
             Otherwise, we could end up calling fread after EOF.  */
          if (feof (stream))
            goto process_partial_block;
        }

      /* Process buffer with BLOCKSIZE bytes.  Note that
         BLOCKSIZE % 64 == 0
       */
      md5_process_block (buffer, BLOCKSIZE, &ctx);
    }

process_partial_block:

  /* Process any remaining bytes.  */
  if (sum > 0)
    md5_process_bytes (buffer, sum, &ctx);

  /* Construct result in desired memory.  */
  md5_finish_ctx (&ctx, resblock);
  free (buffer);
  return 0;
}

#if ! HAVE_OPENSSL_MD5
/* Compute MD5 message digest for LEN bytes beginning at BUFFER.  The
   result is always in little endian byte order, so that a byte-wise
   output yields to the wanted ASCII representation of the message
   digest.  */
void *
md5_buffer (const char *buffer, size_t len, void *resblock)
{
  struct md5_ctx ctx;

  /* Initialize the computation context.  */
  md5_init_ctx (&ctx);

  /* Process whole buffer but last len % 64 bytes.  */
  md5_process_bytes (buffer, len, &ctx);

  /* Put result in desired memory area.  */
  return md5_finish_ctx (&ctx, resblock);
}


void
md5_process_bytes (const void *buffer, size_t len, struct md5_ctx *ctx)
{
  /* When we already have some bits in our internal buffer concatenate
     both inputs first.  */
  if (ctx->buflen != 0)
    {
      size_t left_over = ctx->buflen;
      size_t add = 128 - left_over > len ? len : 128 - left_over;

      memcpy (&((char *) ctx->buffer)[left_over], buffer, add);
      ctx->buflen += add;

      if (ctx->buflen > 64)
        {
          md5_process_block (ctx->buffer, ctx->buflen & ~63, ctx);

          ctx->buflen &= 63;
          /* The regions in the following copy operation cannot overlap.  */
          memcpy (ctx->buffer,
                  &((char *) ctx->buffer)[(left_over + add) & ~63],
                  ctx->buflen);
        }

      buffer = (const char *) buffer + add;
      len -= add;
    }

  /* Process available complete blocks.  */
  if (len >= 64)
    {
#if !_STRING_ARCH_unaligned
# define UNALIGNED_P(p) ((uintptr_t) (p) % alignof (uint32_t) != 0)
      if (UNALIGNED_P (buffer))
        while (len > 64)
          {
            md5_process_block (memcpy (ctx->buffer, buffer, 64), 64, ctx);
            buffer = (const char *) buffer + 64;
            len -= 64;
          }
      else
#endif
        {
          md5_process_block (buffer, len & ~63, ctx);
          buffer = (const char *) buffer + (len & ~63);
          len &= 63;
        }
    }

  /* Move remaining bytes in internal buffer.  */
  if (len > 0)
    {
      size_t left_over = ctx->buflen;

      memcpy (&((char *) ctx->buffer)[left_over], buffer, len);
      left_over += len;
      if (left_over >= 64)
        {
          md5_process_block (ctx->buffer, 64, ctx);
          left_over -= 64;
          memcpy (ctx->buffer, &ctx->buffer[16], left_over);
        }
      ctx->buflen = left_over;
    }
}


/* These are the four functions used in the four steps of the MD5 algorithm
   and defined in the RFC 1321.  The first function is a little bit optimized
   (as found in Colin Plumbs public domain implementation).  */
/* #define FF(b, c, d) ((b & c) | (~b & d)) */
#define FF(b, c, d) (d ^ (b & (c ^ d)))
#define FG(b, c, d) FF (d, b, c)
#define FH(b, c, d) (b ^ c ^ d)
#define FI(b, c, d) (c ^ (b | ~d))

/* Process LEN bytes of BUFFER, accumulating context into CTX.
   It is assumed that LEN % 64 == 0.  */

void
md5_process_block (const void *buffer, size_t len, struct md5_ctx *ctx)
{
  uint32_t correct_words[16];
  const uint32_t *words = buffer;
  size_t nwords = len / sizeof (uint32_t);
  const uint32_t *endp = words + nwords;
  uint32_t A = ctx->A;
  uint32_t B = ctx->B;
  uint32_t C = ctx->C;
  uint32_t D = ctx->D;
  uint32_t lolen = len;

  /* First increment the byte count.  RFC 1321 specifies the possible
     length of the file up to 2^64 bits.  Here we only compute the
     number of bytes.  Do a double word increment.  */
  ctx->total[0] += lolen;
  ctx->total[1] += (len >> 31 >> 1) + (ctx->total[0] < lolen);

  /* Process all bytes in the buffer with 64 bytes in each round of
     the loop.  */
  while (words < endp)
    {
      uint32_t *cwp = correct_words;
      uint32_t A_save = A;
      uint32_t B_save = B;
      uint32_t C_save = C;
      uint32_t D_save = D;

      /* First round: using the given function, the context and a constant
         the next context is computed.  Because the algorithms processing
         unit is a 32-bit word and it is determined to work on words in
         little endian byte order we perhaps have to change the byte order
         before the computation.  To reduce the work for the next steps
         we store the swapped words in the array CORRECT_WORDS.  */

#define OP(a, b, c, d, s, T)                                            \
      do                                                                \
        {                                                               \
          a += FF (b, c, d) + (*cwp++ = SWAP (*words)) + T;             \
          ++words;                                                      \
          CYCLIC (a, s);                                                \
          a += b;                                                       \
        }                                                               \
      while (0)

      /* It is unfortunate that C does not provide an operator for
         cyclic rotation.  Hope the C compiler is smart enough.  */
#define CYCLIC(w, s) (w = (w << s) | (w >> (32 - s)))

      /* Before we start, one word to the strange constants.
         They are defined in RFC 1321 as

         T[i] = (int) (4294967296.0 * fabs (sin (i))), i=1..64

         Here is an equivalent invocation using Perl:

         perl -e 'foreach(1..64){printf "0x%08x\n", int (4294967296 * abs (sin $_))}'
       */

      /* Round 1.  */
      OP (A, B, C, D, 7, 0xd76aa478);
      OP (D, A, B, C, 12, 0xe8c7b756);
      OP (C, D, A, B, 17, 0x242070db);
      OP (B, C, D, A, 22, 0xc1bdceee);
      OP (A, B, C, D, 7, 0xf57c0faf);
      OP (D, A, B, C, 12, 0x4787c62a);
      OP (C, D, A, B, 17, 0xa8304613);
      OP (B, C, D, A, 22, 0xfd469501);
      OP (A, B, C, D, 7, 0x698098d8);
      OP (D, A, B, C, 12, 0x8b44f7af);
      OP (C, D, A, B, 17, 0xffff5bb1);
      OP (B, C, D, A, 22, 0x895cd7be);
      OP (A, B, C, D, 7, 0x6b901122);
      OP (D, A, B, C, 12, 0xfd987193);
      OP (C, D, A, B, 17, 0xa679438e);
      OP (B, C, D, A, 22, 0x49b40821);

      /* For the second to fourth round we have the possibly swapped words
         in CORRECT_WORDS.  Redefine the macro to take an additional first
         argument specifying the function to use.  */
#undef OP
#define OP(f, a, b, c, d, k, s, T)                                      \
      do                                                                \
        {                                                               \
          a += f (b, c, d) + correct_words[k] + T;                      \
          CYCLIC (a, s);                                                \
          a += b;                                                       \
        }                                                               \
      while (0)

      /* Round 2.  */
      OP (FG, A, B, C, D, 1, 5, 0xf61e2562);
      OP (FG, D, A, B, C, 6, 9, 0xc040b340);
      OP (FG, C, D, A, B, 11, 14, 0x265e5a51);
      OP (FG, B, C, D, A, 0, 20, 0xe9b6c7aa);
      OP (FG, A, B, C, D, 5, 5, 0xd62f105d);
      OP (FG, D, A, B, C, 10, 9, 0x02441453);
      OP (FG, C, D, A, B, 15, 14, 0xd8a1e681);
      OP (FG, B, C, D, A, 4, 20, 0xe7d3fbc8);
      OP (FG, A, B, C, D, 9, 5, 0x21e1cde6);
      OP (FG, D, A, B, C, 14, 9, 0xc33707d6);
      OP (FG, C, D, A, B, 3, 14, 0xf4d50d87);
      OP (FG, B, C, D, A, 8, 20, 0x455a14ed);
      OP (FG, A, B, C, D, 13, 5, 0xa9e3e905);
      OP (FG, D, A, B, C, 2, 9, 0xfcefa3f8);
      OP (FG, C, D, A, B, 7, 14, 0x676f02d9);
      OP (FG, B, C, D, A, 12, 20, 0x8d2a4c8a);

      /* Round 3.  */
      OP (FH, A, B, C, D, 5, 4, 0xfffa3942);
      OP (FH, D, A, B, C, 8, 11, 0x8771f681);
      OP (FH, C, D, A, B, 11, 16, 0x6d9d6122);
      OP (FH, B, C, D, A, 14, 23, 0xfde5380c);
      OP (FH, A, B, C, D, 1, 4, 0xa4beea44);
      OP (FH, D, A, B, C, 4, 11, 0x4bdecfa9);
      OP (FH, C, D, A, B, 7, 16, 0xf6bb4b60);
      OP (FH, B, C, D, A, 10, 23, 0xbebfbc70);
      OP (FH, A, B, C, D, 13, 4, 0x289b7ec6);
      OP (FH, D, A, B, C, 0, 11, 0xeaa127fa);
      OP (FH, C, D, A, B, 3, 16, 0xd4ef3085);
      OP (FH, B, C, D, A, 6, 23, 0x04881d05);
      OP (FH, A, B, C, D, 9, 4, 0xd9d4d039);
      OP (FH, D, A, B, C, 12, 11, 0xe6db99e5);
      OP (FH, C, D, A, B, 15, 16, 0x1fa27cf8);
      OP (FH, B, C, D, A, 2, 23, 0xc4ac5665);

      /* Round 4.  */
      OP (FI, A, B, C, D, 0, 6, 0xf4292244);
      OP (FI, D, A, B, C, 7, 10, 0x432aff97);
      OP (FI, C, D, A, B, 14, 15, 0xab9423a7);
      OP (FI, B, C, D, A, 5, 21, 0xfc93a039);
      OP (FI, A, B, C, D, 12, 6, 0x655b59c3);
      OP (FI, D, A, B, C, 3, 10, 0x8f0ccc92);
      OP (FI, C, D, A, B, 10, 15, 0xffeff47d);
      OP (FI, B, C, D, A, 1, 21, 0x85845dd1);
      OP (FI, A, B, C, D, 8, 6, 0x6fa87e4f);
      OP (FI, D, A, B, C, 15, 10, 0xfe2ce6e0);
      OP (FI, C, D, A, B, 6, 15, 0xa3014314);
      OP (FI, B, C, D, A, 13, 21, 0x4e0811a1);
      OP (FI, A, B, C, D, 4, 6, 0xf7537e82);
      OP (FI, D, A, B, C, 11, 10, 0xbd3af235);
      OP (FI, C, D, A, B, 2, 15, 0x2ad7d2bb);
      OP (FI, B, C, D, A, 9, 21, 0xeb86d391);

      /* Add the starting values of the context.  */
      A += A_save;
      B += B_save;
      C += C_save;
      D += D_save;
    }

  /* Put checksum in context given as argument.  */
  ctx->A = A;
  ctx->B = B;
  ctx->C = C;
  ctx->D = D;
}
#endif
